src/main/java/org/monarchinitiative/fhir2hpo/fhir/util/ObservationLoincInfo.java

```diff
@@ -0,0 +1,203 @@
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.Optional;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.hl7.fhir.dstu3.model.CodeableConcept;
+import org.hl7.fhir.dstu3.model.Observation;
+import org.hl7.fhir.dstu3.model.Observation.ObservationComponentComponent;
+import org.hl7.fhir.dstu3.model.Observation.ObservationReferenceRangeComponent;
+import org.hl7.fhir.dstu3.model.Quantity;
+import org.hl7.fhir.dstu3.model.StringType;
+import org.hl7.fhir.dstu3.model.Type;
+import org.hl7.fhir.exceptions.FHIRException;
+import org.monarchinitiative.fhir2hpo.loinc.LoincId;
+import org.monarchinitiative.fhir2hpo.loinc.exception.MismatchedLoincIdException;
+
+/**
+ * This class extracts elements of interest from a FHIR observation for a specific LOINC, regardless of whether it is
+ * in the code section or a component.
+ * 
+ * @author yateam
+ *
+ */
+public class ObservationLoincInfo {
+
+	private static final Logger logger = LogManager.getLogger();
+
+	private String fhirId;
+	private LoincId loincId;
+	private String description;
+	private String date;
+	private String valueDescription;
+	private Optional<CodeableConcept> interpretation = Optional.empty();
+	private Optional<Quantity> valueQuantity = Optional.empty();
+	private Optional<List<ObservationReferenceRangeComponent>> referenceRange = Optional.empty();
+	private Optional<String> valueString = Optional.empty();
+
+	public ObservationLoincInfo(LoincId loincId, Observation observation) throws MismatchedLoincIdException {
+		
+		this.fhirId = observation.getIdElement().getIdPart();
+		this.loincId = loincId;
+		this.date = getDateString(observation);
+		boolean containsLoinc = false;
+		if (ObservationUtil.getCodeSectionLoincIdsOfObservation(observation).contains(loincId)) {
+			containsLoinc = true;
+			setFieldsWithReflection(observation);
+		} else {
+			ObservationComponentComponent component = ObservationUtil.getComponentLoincIdsOfObservation(observation)
+				.get(loincId);
+			if (component != null) {
+				containsLoinc = true;
+				setFieldsWithReflection(component);
+			}
+
+		}
+		
+		if (!containsLoinc) {
+			throw new MismatchedLoincIdException("The observation does not contain the LOINC " + loincId);
+		}
+	}
+
+	private String getDateString(Observation observation) {
+		// TODO: Handle effective period or NPEs?
+		try {
+			if (observation.hasEffectiveDateTimeType()) {
+				return observation.getEffectiveDateTimeType().asStringValue();
+			}
+		} catch (FHIRException e) {
+			// This should not occur since we check existence before getting
+			e.printStackTrace();
+		}
+
+		logger.warn("Could not find a date for the observation.");
+		return "";
+	}
+	
+	/**
+	 * These fields may come from the Observation or the ObservationComponentComponent depending
+	 * on where the LOINC of interest lives.
+	 * @param the observation or the component of it relevant to the LOINC
+	 */
```

Comment:
Rather than parsing this data into Objects and then doing a bunch of reflection to get at the data fields, would it make sense to parse it into a Map?

Reply:
If using a Map isn't possible, I'd suggest adding comments or refactoring this into a few smaller methods with intention-revealing names. It's pretty difficult to follow.

Reply:
I'm not doing any parsing myself - the HAPI-FHIR library is, so I don't think I can turn it into a Map. The Object in question is either an Observation or an ObservationComponentComponent. They are not related in a way where the getters I need are part of a common ancestor. To add to the frustration, some of these getters will throw an exception if the thing you're trying to get is null. With Java being strongly typed, the only options I saw were to implement this method twice - once for each type of object, or to implement it once through reflection. I don't like either, but at least with reflection I don't have to worry about the two methods getting out of sync. If you have suggestions, let me know. Otherwise I'll add some comments to make things clearer.